aten/src/ATen/native/Batching.cpp

```diff
@@ -20,6 +21,73 @@ static bool has_level(const Tensor& self, int64_t level) {
   return it != bdims.end();
 }
 
+// Returns a Tensor with batch dim with level `level` turned into a regular dimension,
+// as well as a logical dim index of where said dimension is in the returned tensor.
+//
+// The reason why we want to return the index of where said dimension is in the returned
+// tensor is because we want to keep track of which dimension used to be the batch
+// dimension so that we can move it to the correct logical dimension specified by
+// `out_dims` in vmap. For example, if we had
+// >>> x = torch.randn(2, 3, 5)
+// >>> vmap(lambda x: x, in_dims=0, out_dims=1)(x)
+// then right when we are about to exit the vmap block, x is a BatchedTensor with a
+// batch dimension at (physical) index 0. Note that the batch dimension doesn't
+// always have to exist at (physical) index 0. When we undo the batch dimension,
+// we want to move it to dimension 1 (as specified by out_dims). So we return the
+// index at which the batch dim appears so that we can move it to the correct place.
+// later down the line.
+static std::pair<Tensor,int64_t> remove_existing_batch_dim(const Tensor& self, int64_t level) {
+  const auto* batched = maybeGetBatched(self);
+  TORCH_INTERNAL_ASSERT(batched != nullptr);
+  auto bdims = batched->bdims();
+  if (bdims.size() == 1) {
+    TORCH_INTERNAL_ASSERT(bdims[0].level() == level);
+    return std::make_pair(batched->value(), bdims[0].dim());
+  }
+  BatchDims new_bdims;
+  int64_t newly_exposed_physical_dim;
```

Comment:
It seems possible for `newly_exposed_physical_dim` to pass through the loop below without being set. I guess there is a bdims invariant which would ensure this is the case. But you could be extra careful by initializing this to -1 and then asserting that it is not -1 at the end.

Reply:
The assert sounds good. Yeah, this function assumes that the level must exist in the bdims.

Reply:
Added assert